src/pymor/tools/random.py

```diff
@@ -2,54 +2,104 @@
 # Copyright pyMOR developers and contributors. All rights reserved.
 # License: BSD 2-Clause License (https://opensource.org/licenses/BSD-2-Clause)
 
+from contextvars import ContextVar
+import inspect
+
 from pymor.core.defaults import defaults
 
 import numpy as np
 
 
-def get_random_state(random_state=None, seed=None):
-    """Returns a |NumPy| :class:`~numpy.random.RandomState`.
-
-    Parameters
-    ----------
-    random_state
-        If specified, this state is returned.
-    seed
-        If specified, the seed to initialize a new random state.
-
-    Returns
-    -------
-    Either the provided, a newly created or the default `RandomState`
-    object.
-    """
-    assert random_state is None or seed is None
-    if random_state is not None:
-        return random_state
-    elif seed is not None:
-        return np.random.RandomState(seed)
-    else:
-        return default_random_state()
+@defaults('seed_seq')
+def init_rng(seed_seq=42):
+    if not isinstance(seed_seq, np.random.SeedSequence):
+        seed_seq = np.random.SeedSequence(seed_seq)
+    # Store a new rng together with seed_seq, as the latter cannot be recovered from the
+    # rng via a public interface (see https://github.com/numpy/numpy/issues/15322).
+    # The first field is a flag to indicate whether the current _rng_state has been consumed
+    # via get_rng. This is a safeguard to detect calls to get_rng in concurrent code
+    # paths.
+    _rng_state.set([False, np.random.default_rng(seed_seq), seed_seq])
+
+
+def get_rng():
+    rng_state = _get_rng_state()
+    rng_state[0] = True
+    _rng_state.set([False] + rng_state[1:])
+    return rng_state[1]
+
+
+class set_rng:
```

Comment:
It looks like you made an effort to make instances of this class reusable. Can you say why?
Also (related?)  `__enter__` is guarded against multiple entry, can you explain the reason here as well?
I do not see how this would actually happen (from same-thread code flow).